### PR TITLE
added composer focus glitch fix

### PR DESCRIPTION
**Pull request by ayazalavi**
### Details
In this PR, I have added a method to maintain focus on the input box by passing its input reference and modal status. This prevents the composer from stealing focus away from that input box.

### Fixed Issues
1. $ https://github.com/Expensify/App/issues/26527
2. PROPOSAL: https://github.com/Expensify/App/issues/26527#issuecomment-1708380347

### Tests

1. Open the app
2. Open any report
3. Edit any message
4. Click CTRL+J to open keyboard shortcut or click on delete icon of some other chat to open delete modal
5. now close the modal
6. Edit message box should not glitch as it works currently on ESC from CTRL+K or CTRL+SHIFT+K shortcuts

- [x] Verify that no errors appear in the JS console

### Offline tests
same as online

### QA Steps
same as above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/767439/802fc37f-163e-471c-b5ee-845f70a18a2c

https://github.com/Expensify/App/assets/767439/d7ab2394-46cc-4ba4-b3c0-e9c5f8860913

</details>

<details>
<summary>Mobile Web - Chrome </summary>

https://github.com/Expensify/App/assets/767439/8c534140-7939-4c6f-8e45-88d937bde925

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/767439/57bc0b51-0ff9-414d-863a-edb61b44c030

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/767439/a7bfe59a-f7b9-4506-a5b2-9af250d76bf3

</details>

<details>
<summary>iOS</summary>

Should work same as before

</details>

<details>
<summary>Android</summary>

Should work same as before

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionItemMessageEdit.js [hunk 1]
```diff
@@ -124,6 +129,39 @@ function ReportActionItemMessageEdit(props) {
         isFocusedRef.current = isFocused;
     }, [isFocused]);
 
+    useEffect(() => {
+        if (Platform.OS === 'web' && !Browser.isMobile()) {
```
Comment (narefyev91) [inline src/pages/home/report/ReportActionItemMessageEdit.js:133]: we should not use such checks - you need to create a separate files for handle web and other platforms

File: src/pages/home/report/ReportActionItemMessageEdit.js [hunk 2]
```diff
@@ -124,6 +129,39 @@ function ReportActionItemMessageEdit(props) {
...
+            InputFocus.composerFocusKeepFocusOn(textInputRef.current, isFocused, modal, onyxFocused);
+        }
+        return () => {};
```
Comment (narefyev91) [inline src/pages/home/report/ReportActionItemMessageEdit.js:136]: i think this is not needed

**Comment by narefyev91** [comment]
@ayazalavi hey! Seems issue with jumping borders are not fixed on desktop

https://github.com/Expensify/App/assets/28352309/62f9e36b-fed1-43ef-a8d9-cb5ec3824460


https://github.com/Expensify/App/assets/28352309/9d0012d3-db3e-43a5-8831-9a0ef58fc667

Also i think default behaviour now changed - when i open a comment for edit -> open shortcuts modal -> and close it:
Expected behaviour - main composer should be focused
Actual result - edit field is focused

Also when i after it click on main composer -> open shortcuts modal -> and close it:
Expected behaviour - main composer should be focused
Actual result - edit field is focused

https://github.com/Expensify/App/assets/28352309/a5df83fc-4599-45e8-b0a3-db77bb602528

**Comment by ayazalavi** [comment]
@narefyev91 I have uploaded` index.desktop.ts` so it should work fine on desktop as well now.

I am using same technique as on mobile to put focus back to last input field we had focus on. 

https://github.com/Expensify/App/assets/767439/161847af-1ac2-4851-a06d-e4d91f9e3b1e

Other than that if user explicitly set focus on compose box then focus will stay there. This matches with mobile web as well.

https://github.com/Expensify/App/assets/767439/bef2ce02-9684-4e78-a625-8645702ad914

**Comment by ayazalavi** [comment]
@narefyev91 have you been able to check PR? any more changes you want?

**Comment by narefyev91** [comment]
> @narefyev91 have you been able to check PR? any more changes you want?

in progress

**Comment by narefyev91** [comment]
@ayazalavi one think i noticed comparing how staging and local changes are working
When we go back from screen (closing right panel) we should focus main composer - and not any focus edit messages
Staging:

https://github.com/Expensify/App/assets/28352309/3550fd78-3125-437f-8393-98093a1c7d44

Locally:

https://github.com/Expensify/App/assets/28352309/845e53c5-5faa-4d53-881f-72eb22605829

**Comment by narefyev91** [comment]
Hmmm interesting we have some different behaviours between web, mobile web and native
Let's for now leave as it is - i will clarify

**Comment by ayazalavi** [comment]
@narefyev91, yes, all three platforms have different implementations. React Native Modal, by default, puts focus back on the last input that was focused. So, I am trying to maintain that behavior while avoiding the glitch. Secondly, if the edit message is opened, then it makes perfect sense to put the focus back on it once the modal gets closed unless the user explicitly clicks on the main composer box.

If it is possible to merge this PR by tomorrow, then I'll be really grateful to get the bonus.

**Comment by narefyev91** [comment]
@ayazalavi I tested on web and desktop - looks good
But seems on latest main we have some issues in logic
When user go back from report - and come back - we focus on edit message composer, when we click to open user details - and go back - we focus main composer - to make it consistently we need to focus edit message composer 

https://github.com/Expensify/App/assets/28352309/015ce68a-a1d5-4918-a371-05d720115ebc

**Comment by ayazalavi** [comment]
> @ayazalavi I tested on web and desktop - looks good But seems on latest main we have some issues in logic When user go back from report - and come back - we focus on edit message composer, when we click to open user details - and go back - we focus main composer - to make it consistently we need to focus edit message composer
> 
>  Screen.Recording.2023-10-11.at.16.25.53.mov

React native modal puts focus back to the last input that was focused before modal was opened. So as soon as user details page closes then basically focus will go back automatically to the edit input box. We cannot put focus back to main composer in this case.

**Comment by narefyev91** [comment]
> > @ayazalavi I tested on web and desktop - looks good But seems on latest main we have some issues in logic When user go back from report - and come back - we focus on edit message composer, when we click to open user details - and go back - we focus main composer - to make it consistently we need to focus edit message composer
> > Screen.Recording.2023-10-11.at.16.25.53.mov
> 
> React native modal puts focus back to the last input that was focused before modal was opened. So as soon as user details page closes then basically focus will go back automatically to the edit input box. We cannot put focus back to main composer in this case.

@ayazalavi Yes - it should focus edit composer - but as you can see on the video - it focused main composer

**Comment by ayazalavi** [comment]
> > > @ayazalavi I tested on web and desktop - looks good But seems on latest main we have some issues in logic When user go back from report - and come back - we focus on edit message composer, when we click to open user details - and go back - we focus main composer - to make it consistently we need to focus edit message composer
> > > Screen.Recording.2023-10-11.at.16.25.53.mov
> > 
> > 
> > React native modal puts focus back to the last input that was focused before modal was opened. So as soon as user details page closes then basically focus will go back automatically to the edit input box. We cannot put focus back to main composer in this case.
> 
> @ayazalavi Yes - it should focus edit composer - but as you can see on the video - it focused main composer when coming back from personal details screen

okay got it, I have added changes.

**Comment by narefyev91** [comment]
## Reviewer Checklist
- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/28352309/6b31283e-27b8-47fa-b4d7-abefd674a1e9




</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/28352309/fc7a78cb-7e92-4f6e-b7e6-0e643ba045d5




</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/28352309/5f2fc037-256e-46a5-8573-377c7de5e0f9




</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/28352309/ed6320a4-ed10-45ad-b7f1-8a808bba009a




</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/28352309/08ad6066-42ca-44cf-bb72-5938aa5f81c5




</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/28352309/f2ee0c08-3661-4fde-9a8c-b7042e133b2e




</details>

**Review by narefyev91 — APPROVED** [review]
LGTM! Working fine and resolved the main issue
🎀 👀 🎀 C+ reviewed

**Comment by ayazalavi** [comment]
@amyevans, I hope you're well. Could you please review my pull request as soon as possible? Your prompt feedback would be highly appreciated. Thank you in advance for your quick attention.

**Comment by amyevans** [comment]
Sorry, it's on my to do list, I just have several higher priority items I need to get through first. Aiming for later today or tomorrow at the latest.

**Review by amyevans — APPROVED** [review]
Looks good, thanks guys!

**Comment by ayazalavi** [comment]
@narefyev91 Can we reopen this PR and send update with the performance fix?

**Comment by ayazalavi** [comment]
Thanks @amyevans for clarifying the issue.

**Pull request merged**